cogent/src/Cogent/Glue.hs

```diff
@@ -837,25 +837,57 @@ readEntryFuncs :: [SF.TopLevel TC.DepType TC.TypedPatn TC.TypedExpr]
-  = foldM (\m ln -> readEntryFunc ln >>= \(fn,inst) -> return $ updateFunMono m fn inst) M.empty lns
+  = foldM run (Just M.empty) lns
   where
+    run md ln =
+      case md of
+        Nothing -> return Nothing
+        Just m  -> do
+          f <- readEntryFunc ln
+          case f of
+            Nothing -> return Nothing
+            Just (fn, inst) ->
+              return $ Just $ updateFunMono m fn inst
+
     updateFunMono m fn ([],[]) = M.insertWith (\_ entry -> entry) fn M.empty m
     updateFunMono m fn inst    = M.insertWith (\_ entry -> M.insertWith (flip const) inst (M.size entry) entry) fn (M.singleton inst 0) m
 
     -- Each string is a line in the @--entry-funcs@ file.
-    readEntryFunc :: String -> IO (FunName, MN.Instance VarName)
+    readEntryFunc :: String -> IO (Maybe (FunName, MN.Instance VarName))
     readEntryFunc ln = do
       er <- runExceptT $ flip evalStateT (mkGlState [] tcState dsState mempty (mempty, mempty) undefined) $
               flip runReaderT (FileState "--entry-funcs file") $ do
                 (fnName, targs) <- genFuncId ln noLoc
-                inst <- forM targs $ \mt ->
-                          case mt of
-                            Nothing -> throwError "No wildcard allowed."
-                            Just t  -> flip runReaderT (DefnState Vec.Nil []) $
-                                         flip runReaderT (MonoState (([], []), Nothing))
-                                                         (lift . tcType >=> lift . desugarType >=> monoType $ t)
-                return (fnName, inst)
-      case er of Left s  -> putStrLn s >> return (ln, ([], []))
-                 Right r -> return $ (fst r,) (snd r, [])
+                let nargs = SF.numTypeVars $ 
+                      case find isFnName tced of
+                          Just f  -> f
+                          Nothing -> __impossible "Could not find function in top level declarations"
+                if nargs /= L.length targs then do
+                  throwError ("Number of type arguments for function " ++
+                              fnName ++ " (" ++ show nargs ++
+                              ") does not match amount specified in --entry-funcs file (" ++
+                              show (L.length targs) ++ ").\n" ++
+                              optMsg (nargs > L.length targs))
+                else do
+                  inst <- forM targs $ \mt ->
+                            case mt of
+                              Nothing -> throwError "Use of wildcard disallowed in --entry-funcs file"
+                              Just t  -> flip runReaderT (DefnState Vec.Nil []) $
+                                          flip runReaderT (MonoState (([], []), Nothing))
+                                                          (lift . tcType >=> lift . desugarType >=> monoType $ t)
+                  return (fnName, inst)
+      case er of Left s  -> putStrLn ("\nError: " ++ s) >> return Nothing
```

Comment:
it should go to stderr.

Reply:
do we have a standard way of doing this?

Reply:
No, I just used ` hPutStrLn stderr`.